File: src/utils_fs/Diagnostics.fs
#light

namespace Utils

open System;
open System.Diagnostics;

module public Diagnostics = 
    let public Profile action = 
        let stopwatch = new Stopwatch()
        stopwatch.Start()

        let result = action()

        stopwatch.Stop()

        let stackTrace = new StackTrace(false)
        let stackFrame = stackTrace.GetFrame(1);
        
        (result, stopwatch, stackFrame)

    let public Out (result, (stopwatch:Stopwatch), (stackFrame:StackFrame)) = 
        Console.WriteLine("msec: {0} in {1}",
                    stopwatch.Elapsed.TotalMilliseconds,
                    stackFrame.GetMethod());

        Console.WriteLine("Die Antwort auf alle Fragen: {0}", box result)
        
    let public ProfileOut action = 
        Profile action |> Out
        

File: src/utils_cs/Diagnostics.cs
using System;

namespace Utils
{
	using System.Diagnostics;
	using System.Linq.Expressions;

	public static class Diagnostics
	{
		public static void Profile(Func<int> @function)
		{
			var stopwatch = new Stopwatch();
			stopwatch.Start();

			var result = @function();

			stopwatch.Stop();

			var stackTrace = new StackTrace(false);
			Console.WriteLine("msec: {0} in {1}",
							   stopwatch.Elapsed.TotalMilliseconds,
							   stackTrace.GetFrame(1).GetMethod());

			Console.WriteLine("Die Antwort auf alle Fragen: {0}",
							   result);
		}

		public static void ProfileExpression(Expression<Func<int>> @function)
		{
			var stopwatch = new Stopwatch();
			stopwatch.Start();

			var result = @function.Compile()();

			stopwatch.Stop();

			var stackTrace = new StackTrace(false);
			Console.WriteLine("msec: {0} in {1}",
							   stopwatch.Elapsed.TotalMilliseconds,
							   stackTrace.GetFrame(1).GetMethod());

			Console.WriteLine("Die Antwort auf alle Fragen: {0}",
							   result);
		}
	}
}
